firedrake/meshadapt.py

```diff
@@ -0,0 +1,500 @@
+    """
+    Class for driving metric-based mesh adaptation.
+    """
+
+    @PETSc.Log.EventDecorator()
+    def __init__(self, mesh, metric):
+        """
+        :param mesh: :class:`~firedrake.mesh.MeshGeometry` to be adapted
+        :param metric: :class:`.RiemannianMetric` to use for the adaptation
+        """
+        if metric._mesh is not mesh:
+            raise ValueError("The mesh associated with the metric is inconsistent")
+        if isinstance(mesh.topology, fmesh.ExtrudedMeshTopology):
+            raise NotImplementedError("Cannot adapt extruded meshes")
+        coord_fe = mesh.coordinates.ufl_element()
+        if (coord_fe.family(), coord_fe.degree()) != ("Lagrange", 1):
+            raise NotImplementedError(f"Mesh coordinates must be P1, not {coord_fe}")
+        assert isinstance(metric, RiemannianMetric)
+        super().__init__(mesh)
+        self.metric = metric
+
+    @futils.cached_property
+    @PETSc.Log.EventDecorator()
+    def adapted_mesh(self):
+        """
+        Adapt the mesh with respect to the provided metric.
+
+        :return: a new :class:`~firedrake.mesh.MeshGeometry`.
+        """
+        self.metric.enforce_spd(restrict_sizes=True, restrict_anisotropy=True)
+        v = self.metric._create_from_array(self.metric.dat.data_with_halos)
+        newplex = self.metric._plex.adaptMetric(v, "Face Sets", "Cell Sets")
+        v.destroy()
+        return fmesh.Mesh(newplex, distribution_parameters={"partition": False})
+
+    @PETSc.Log.EventDecorator()
+    def project(self, f):
+        """
+        Project a :class:`.Function` into the corresponding :class:`.FunctionSpace`
+        defined on the adapted mesh using supermeshing.
+
+        :param: the scalar :class:`.Function` on the initial mesh
+        :return: its projection onto the adapted mesh
+        """
+        fs = ffs.FunctionSpace(self.adapted_mesh, f.ufl_element())
+        return firedrake.project(f, fs)
+
+    @PETSc.Log.EventDecorator()
+    def interpolate(self, f):
+        """
+        Interpolate a :class:`.Function` into the corresponding :class:`.FunctionSpace`
+        defined on the adapted mesh.
+
+        :param: the scalar :class:`.Function` on the initial mesh
+        :return: its interpolation onto the adapted mesh
+        """
+        raise NotImplementedError(
+            "Consistent interpolation has not yet been implemented in parallel"
+        )  # TODO
+
+
+def adapt(mesh, *metrics):
+    r"""
+    Adapt a mesh with respect to a metric and some adaptor parameters.
+
+    If multiple metrics are provided, then they are intersected.
+
+    :param mesh: :class:`~firedrake.mesh.MeshGeometry` to be adapted.
+    :param metrics: list of :class:`.RiemannianMetric`\s
+    :return: a new :class:`~firedrake.mesh.MeshGeometry`.
+    """
+    metric = metrics[0]
+    if len(metrics) > 1:
+        metric.intersect(*metrics[1:])
+    adaptor = MetricBasedAdaptor(mesh, metric)
+    return adaptor.adapted_mesh
```

Comment:
Would it be possible to hide `RiemannianMetric` from users? I would expect some interface like:
```
mesh = ...
V = FunctionSpace(mesh, ...)
f = Function(V)  # some scalar field
h = Function(V).interpolate(...)  # another field
g0 = compute_hessian_from_scalar_field(f)  # a utility function; returns a tensor P1 function
g1 = ...  # define another tensor P1 function if we want
amsh = AdaptedMesh(mesh, [g0, g1], average=False, normalise=False, name="my_adapted_mesh")
Vadapt = FunctionSpace(amsh, ...)
hadapt = Function(Vadapt).interpolate(h)  # Use the regular interface for interpolation (and projection).
```
In the last line we know if we are interpolating across parent mesh-adapted mesh if we let `amsh` carry the adaptor object as attribute (e.g., `amsh.adapted_adaptor`, where `adapted_` is a prefix).

I cannot think through all possible use cases, so would like to hear what others think.

Reply:
I don't think it's a very nice pattern to have a constructor (or function masquerading as a constructor) that takes another instance of the same class as a constructor parameter. I think `adapt(mesh, ...)` is a better interface than `AdaptedMesh(mesh, ...)`.

Reply:
I think `AdaptedMesh` is more consistent with other similar functions. We already have `ExtrudedMesh`, `VertexOnlyMesh`, and now `RelabeledMesh`, which have similar interface.

Reply:
I think the difference I see between this and an `ExtrudedMesh` or `VertexOnlyMesh` is that those functions need to remember their initial meshes. In other words it makes sense for them to retain `mesh.base_mesh` or `mesh.parent_mesh` properties. Is it ever the case that we need `adapted_mesh.original_mesh`? Arguably `RelabeledMesh` should be `relabel(mesh, ...)` as it (AIUI) fits the same paradigm.

Reply:
I think it's fair to store parent mesh as long as the derived meshes are related to the parent mesh; in this specific case, adapted meshes represent the same domain as the parent mesh, which is important.

Reply:
I think, when we discussed this many years ago, we had a concept of a `MeshAdaptor` as joe introduces. The idea being you need both meshes lying around until such time as you've transferred all the fields you need to. The `adaptor` object can be used for holding both meshes and the (potentially expensive) supermesh projection mass-matrices.

But maybe that is no longer necessary.

Reply:
> Would it be possible to hide `RiemannianMetric` from users?

I think `RiemannianMetric` needs to be available to users. Many of the manipulations that people do on metrics (especially for goal-oriented adaptation) need the methods provided here at several steps and cannot be straightforwardly done using methods of `Function` alone.

Reply:
> The `adaptor` object can be used for holding both meshes and the (potentially expensive) supermesh projection mass-matrices.

Good point on mass matrices. I guess the way to do that would be to associate a `Projector` with the `MetricBasedAdaptor`, rather than just calling `project`?